### PR TITLE
RFC: Headless core components & imperative support

### PR DESCRIPTION
**Pull request by andrewdoro**
@steven-tey Really cool project, I really love it 😍

The whole purpose of this PR is to **_radixify_** the `core` package and make it easier to add custom behaviour and most essentially custom styling. It decouples the styling from tailwind completely. It also enables custom bubble & slash command components. (#128,#135, #134)

The code is still not perfect, I just started this a few hours ago and I need to sleep. The structure and the naming conventions are still shady (also the TS types), but overall the direction is to have something headless like Radix-UI.

I also want to add an example with different styling methods (CSS Modules, PandaCSS). 

## Main changes:

- Refactored `editor/index.tsx` to use `EditorProvider` and exposed `useCurrentEditor` to user
- Created `createDefaultExtensions` fn util for default styling
- Added `BubbleMenuItem` unstyled component with polymorphic `asChild` being used for commands in the bubble

  ``` tsx
   const item = {
        name: "bold",
        isActive: (editor) => editor.isActive("bold"),
        command: (editor) => editor.chain().focus().toggleBold().run(),
        icon: BoldIcon,
      } as BubbleMenuItemType
  
   <BubbleMenuItem
            command={item.command}
            className="p-2 text-stone-600 hover:bg-stone-100 active:bg-stone-200"
            type="button"
          >
           {...}
  </BubbleMenuItem>
  ```
- Added `BubbleMenu` wrapper over `@tiptap/react/BubbleMenu`
- Exposes **SlashCommand** as fn
``` tsx
const SlashCommand = (list: ComponentType<any>) =>
  Command.configure({
    suggestion: {
      items: getSuggestionItems,
      render: renderItems(list),
    },
  });
 ```
 - Migrated example app to use the new approach (_maybe some users would still want the default tailwind behavirour, this is just for testing now_)
 - The consumer app codes becomes something like this where children of `BubbleMenu` use `BubbleMenuItem` internally
 ``` tsx
  <NovelEditor>
        <BubbleMenu >
          <NodeSelector />
          <LinkSelector />
          <TextBubbles />
          <ColorSelector />
        </BubbleMenu>
  </NovelEditor>
```

### CODE REVIEW

**Comment by andrewdoro** [comment]
I will make some changes to this to be non-destructive

**Comment by andrewdoro** [comment]
## Update 
@steven-tey 

- Kept `core` package until a decision is made (nothing changed here)
- Added `@novel/headless` package which exposes unstyled components and utility functions. This package is unstyled and components can be used to build an Editor. I think we should publish it.
- `NodeSelector`, `LinkSelector` and `ColorSelector` are not part of the `headless` because they can easily be created with the new `EditorBubbleItem`. This also helps in decoupling the selectors from Radix.
- Added `@novel/tailwind` package which uses the headless package and exports a styled Editor
- Currently the `@novel/tailwind` package is used more for showing how to build your own editor with custom styled bubble and command list components. If we publish this package we should export individual  styled components and then compose them in the consumer app.

If we require full flexibility we should also look over implementing a CLI that copies the headless and tailwind package giving user full control, similar to https://github.com/shadcn-ui/ui.

## Editor anatomy

``` tsx
import {
  EditorProvider,
  EditorBubble,
  EditorBubbleItem,
  EditorCommandItem,
  EditorCommandList,
  CommandListProps,
  CommandListItemProps,
  createEditorSlashCommand,
} from "@novel/headless";
import { Popover } from "@radix-ui/react-popover";

const Selector = () => {
  return (
    <Popover>
      <Popover.Trigger />
      <Popover.Content>
        <EditorBubbleItem />
        <EditorBubbleItem />
        <EditorBubbleItem />
      </Popover.Content>
    </Popover>
  );
};

const CommandItem = (props: CommandListItemProps) => (
  <EditorCommandItem {...props}></EditorCommandItem>
);
const CommandList = (props: CommandListProps) => (
  <EditorCommandList {...props} Element={CommandItem} />
);

const Editor = () => {
  return (
    <EditorProvider extensions={[createEditorSlashCommand(CommandList)]}>
      <EditorBubble>
        <EditorBubbleItem />
        <EditorBubbleItem />
        <EditorBubbleItem />
        <Selector />
      </EditorBubble>
    </EditorProvider>
  );
};

export default Anatomy;

File: packages/headless/src/extensions/index.tsx [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    HTMLAttributes: {
+      class: updatedImage,
+    },
+  }),
+  Placeholder.configure({
+    placeholder: ({ node }) => {
+      if (node.type.name === "heading") {
+        return `Heading ${node.attrs.level}`;
+      }
+      return "Press '/' for commands, or '++' for AI autocomplete...";
```
Comment (jeloi) [inline packages/headless/src/extensions/index.tsx:145]: could we make this a param thats passed in? alternative is to just supporting override extensions entirely to configure a custom one, but could be too heavy handed.

**Comment by JimmyLv** [comment]
any update? 😁

**Comment by gediminastub** [comment]
Maybe any update? Looking forward to this very much! Thanks!

**Comment by George-field** [comment]
Any updates on this? happy to help out if not 😄

**Comment by FranciscoMoretti** [comment]
Give this PR a cookie 🍪

**Comment by andrewdoro** [comment]
Update: Just wanted to share an update related to this PR. I've spoken with @steven-tey and I'll join the project as a maintainer. I'll first start by finishing this PR and have it merged.

**Pull request merged**